inspector: introduce inspector/promises API

ErickWendel opened the pull request:
This PR introduces the new `inspector/promises` API  with the `session.post` function.


Before with Callbacks API:
```js
const { Session } = require('inspector');
const session = new Session();
session.connect();

session.post('Profiler.enable', () => {
  session.post('Profiler.start', () => {
    session.post('Profiler.stop', (err, { profile }) => {
        fs.writeFileSync('./profile.cpuprofile', JSON.stringify(profile));
       session.disconnect();
    });
  });
});
```
After with Promises:

```js
const { Session } = require('inspector/promises');
const session = new Session();
session.connect();

await session.post('Profiler.enable');
await session.post('Profiler.start');
 
const { profile } = await session.post('Profiler.stop')
fs.writeFileSync('./profile.cpuprofile', JSON.stringify(profile));

session.disconnect();
```

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

MoLow commented:
Perhaps we should expose `node:inspect/promises`? Just for the sake of consistency

devsnek commented on lib/inspector.js:
```diff
@@ -128,13 +129,26 @@ class Session extends EventEmitter {
     }
-    this[connectionSymbol].dispatch(JSONStringify(message));
+
+    const event = new EventEmitter();
+    const kOnFinish = Symbol('kOnFinish');
+    const promisifiedResult = promisify(event.once.bind(event))(kOnFinish);
+    const cb = (err, res) => event.emit(kOnFinish, err, res);
+    configureCbAndDispatch(cb);
+
+    return promisifiedResult;
```
i don't think we need to create all this event and promisify stuff in the middle? 
```suggestion
    return new Promise((resolve, reject) => {
      configureCbAndDispatch((err, res) => {
        if (err) reject(err);
        else resolve(res);
      });
    });
```

ErickWendel replied:
I loved it! I try doing with much more lines as I thought I’d be better for maintenance but for performance reasons I’d go with what you’ve suggested 🤩

already committed it!

ErickWendel commented:
> Perhaps we should expose `node:inspect/promises`? Just for the sake of consistency

I thought about it, but the only function that makes sense using promises is the post function. When looking at the module there’re a lot of dependencies needed to be rewritten in case we wanna expose is as /promises

ovflowd commented on lib/inspector.js:
```diff
@@ -108,7 +109,7 @@ class Session extends EventEmitter {
    * @param {Record<unknown, unknown>} [params]
    * @param {Function} [callback]
-   * @returns {void}
+   * @returns {void | Promise}
```
I believe it might make sense to add overloading types, by using jsdoc `@type` with something inline of:

```jsdoc
/**
 * @type {{
 * (method:string, params:Record<unknown, unknown>, callback:Function) => void;
 * (method:string, params:Record<unknown) => Promise;
 * }}
 */
```

But I'm honestly not sure how this was done for other functions that support overloading. I'm just curious about the TypeScript counterpart for these types when on `types/node`.

**Note.:** My comment is just a nitpick, not an actual blocker. Cheers!

ovflowd replied:
There's also the other way of inferencing within jsdoc

```jsdoc
@returns {callback extends Function ? void : Promise}
```

ljharb commented on doc/api/inspector.md:
```diff
@@ -183,15 +183,25 @@ changes:
 a response is received. `callback` is a function that accepts two optional
 arguments: error and message-specific result.
 
+If the `callback` is not passed, then it'll return a Promise result.
```
maybe this should be done via util.promisify instead of implicit behavior?

ErickWendel replied:
Well, as a lot of other functions were converted to Promises I thought this might be something good to the user land. WDYT?

ljharb replied:
Are there a lot that do it in this way, with an optional callback?

ljharb replied:
To be clear, what I’m suggesting is adding a custom promisify implementation if needed, and *not* returning a promise from the actual function - so that users can `util.promisify(session.post)` explicitly.

ErickWendel replied:
hey! I just changed to /promises api. Would you check it now?

Trott commented:
@nodejs/inspector

benjamingr commented:
Overall big +1 on the idea

> When looking at the module there’re a lot of dependencies needed to be rewritten in case we wanna expose it as /promises

Why? It can even live in the same file it'd just have to be exported as `/promises`.

I'm not sure why this should be the one and only API that works with "omit callback to get promise back" in Node.js (this specific approach was rejected for fs for example). Namely this would mean that someone forgetting to pass a callback would now stop getting an error and would get an unhandled rejection if the `.post` failed instead since they might be unaware of the new return value.

benjamingr commented on doc/api/inspector.md:
```diff
@@ -183,15 +183,25 @@ changes:
 // Output: { type: 'number', value: 4, description: '4' }
+
+// or for promise
+session.post('Runtime.evaluate', { expression: '2 + 2' })
```
prefer async/await in examples?

ErickWendel replied:
fixed!

MoLow commented on lib/inspector/promises.js:
```diff
@@ -0,0 +1,22 @@
+}
+
+module.exports = {
+  Session: SessionPromises
```
```suggestion
  Session: SessionPromises
  ...inspectorBase
```

ErickWendel replied:
thanks! done!

ErickWendel commented:
> Overall big +1 on the idea
> 
> > When looking at the module there’re a lot of dependencies needed to be rewritten in case we wanna expose it as /promises
> 
> Why? It can even live in the same file it'd just have to be exported as `/promises`.
> 
> I'm not sure why this should be the one and only API that works with "omit callback to get promise back" in Node.js (this specific approach was rejected for fs for example). Namely this would mean that someone forgetting to pass a callback would now stop getting an error and would get an unhandled rejection if the `.post` failed instead since they might be unaware of the new return value.

I struggled a bit but I was able to do it o/

ErickWendel commented:
I think I'll need a deep review of the docs as I'm not sure if I'm on the right path. I copied and pasted the whole API, changed the topics, and added sections for callbacks and promises APIs following the example seeing on [fs](https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#callback-example)

benjamingr commented on doc/api/inspector.md:
````diff
@@ -11,86 +11,205 @@ inspector.
 It can be accessed using:
 
-```js
+To use the promise-based APIs:
````
This paragraph doesn't really make sense with:

```
It can be accessed using:

To use the promise-based APIs:
```

Maybe rephrase this whole area?

ErickWendel replied:
I thought it's easier removing them and add just an `or` there. What do you think? (just committed the change as well)

benjamingr reviewed (commented):
Good progress!

ljharb commented on lib/inspector/promises.js:
```diff
@@ -0,0 +1,21 @@
+const inspector = require('inspector');
+const { promisify } = require('util');
+class Session extends inspector.Session {
+
+  /**
+   * Posts a message to the inspector back-end.
+   * @param {string} method
+   * @param {Record<unknown, unknown>} [params]
+   * @returns {Promise}
+   */
+  async post(method, params) {
+    return promisify(super.post.bind(this))(method, params);
```
```suggestion
  #post = promisify(super.post.bind(this));

  /**
   * Posts a message to the inspector back-end.
   * @param {string} method
   * @param {Record<unknown, unknown>} [params]
   * @returns {Promise}
   */
  async post(method, params) {
    return this.#post(method, params);
```
this way, the original method is looked up once, the bound method is created once, and cached privately on the instance.

ErickWendel replied:
uhuu nice! I loved it! Thanks

MoLow commented on doc/api/inspector.md:
````diff
@@ -11,86 +11,203 @@ inspector.
 ```
 
-## `inspector.close()`
+## Promises API
````
```suggestion
## Promises API

<!-- YAML
added: REPLACEME
-->
```

ljharb commented:
Can you confirm that this will be requireable/importable as BOTH `inspector/promises` and `node:inspector/promises`?

ErickWendel commented:
> Can you confirm that this will be requireable/importable as BOTH `inspector/promises` and `node:inspector/promises`?

I'm not sure if this is managed by the module itself. the `node:` prefix will look for the native modules (as far as I know) but I've tested here and both of them are working fine

ovflowd commented on lib/inspector/promises.js:
```diff
@@ -0,0 +1,21 @@
+   * Posts a message to the inspector back-end.
+   * @param {string} method
+   * @param {Record<unknown, unknown>} [params]
+   * @returns {Promise}
```
Is there any specific type for the Promise response? Is the type somehow related to the `params` param? (Can we maybe give a different name to the param? As a param called `params` could lead to some confusion?

ovflowd commented on doc/api/inspector.md:
```diff
@@ -11,86 +11,207 @@ inspector.
+try {
+  const session = new Session();
+  session.connect();
+  const result = await session.post('Runtime.evaluate', { expression: '2 + 2' });
```
**Question:** would it be worth mentioning that the way internally it works is that instead of passing a callback by using promises, the "result" of the `session.post` is still a deferred callback? 

(I'm assuming that `Promise.all([s.post(1), s.post(2), s.post(3)]` still entails that these promises not necessarily will be resolved in that order as the execution to them is deferred? When I mean not in that order, I'm referring to that post(1) might resolve after post(3))

That might be worth mentioning, as you're not blocking your execution thread when you're using callbacks. Still, there will be a difference in how your execution thread will behave by either using Promises or async/await.

**Note.:** Please ignore if it doesn't apply to this scenario; I'm assuming the internals of the inspector API as I haven't _really_ looked deep into its inner workings.

ErickWendel replied:
Well, honestly I haven't seen a problem while promisifying a function and resulting in a wrong order. I did the test case just in case. Would you check it?

ovflowd replied:
> Well, honestly I haven't seen a problem while promisifying a function and resulting in a wrong order. I did the test case just in case. Would you check it?

Oh, I think you got me wrong the result order should be still the same, I was just referring that the order of resolution might not be the same.

ovflowd commented on test/parallel/test-inspector-promises.js:
```diff
@@ -0,0 +1,44 @@
+    const session = new inspector.Session();
+    session.connect();
+
+    await session.post('Profiler.enable');
```
Could we also add some tests with just plain Promises and some modern Promise utilities such as `Promise.all, allSettled` just curious to cover these scenarios to create an expectancy of how it should behave. (I honestly don't think they wouldn't be respected, but sometimes better have more tests than lesser ;D)

ovflowd reviewed (commented):
Honestly, super cool changes within the Docs. Thank you for these cool contributions ✨

ovflowd commented:
> I think I'll need a deep review of the docs as I'm not sure if I'm on the right path. I copied and pasted the whole API, changed the topics, and added sections for callbacks and promises APIs following the example seeing on [fs](https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#callback-example)

You can build the docs locally, there are some utilities within [here](https://github.com/nodejs/node/tree/main/tools/doc), and there should be a flag to the main `Makefile` of the node to also build the docs. If you could add a screenshot of how they look with your changes, that'd be great, as it would save some time of the fellow contributors from building it locally 👀

ErickWendel commented:
> > I think I'll need a deep review of the docs as I'm not sure if I'm on the right path. I copied and pasted the whole API, changed the topics, and added sections for callbacks and promises APIs following the example seeing on [fs](https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#callback-example)
> 
> You can build the docs locally, there are some utilities within [here](https://github.com/nodejs/node/tree/main/tools/doc), and there should be a flag to the main `Makefile` of the node to also build the docs. If you could add a screenshot of how they look with your changes, that'd be great, as it would save some time of the fellow contributors from building it locally 👀

uhh I didn't know that! Gonna check it right now

MoLow commented:
@ErickWendel you can also download the docs from this job: https://github.com/nodejs/node/actions/runs/2919367653

ljharb reviewed (commented):
LGTM (i can't stamp because it would show up as green, but i'm not a collab so it shouldn't be considered an approval)

aduh95 commented on doc/api/inspector.md:
````diff
@@ -11,86 +11,207 @@ inspector.
+Posts a message to the inspector back-end.
+
+```mjs
+import { Session } from 'inspector/promises';
````
```suggestion
import { Session } from 'node:inspector/promises';
```

aduh95 commented on lib/inspector/promises.js:
```diff
@@ -0,0 +1,21 @@
+
+const inspector = require('inspector');
+const { promisify } = require('util');
+class Session extends inspector.Session {
+  #post = promisify(super.post.bind(this));
+  /**
+   * Posts a message to the inspector back-end.
+   * @param {string} method
+   * @param {Record<unknown, unknown>} [params]
+   * @returns {Promise}
+   */
+  async post(method, params) {
+    return this.#post(method, params);
+  }
+}
```
```suggestion

class Session extends inspector.Session {}

Session.prototype.post = promisify(inspector.Session.prototype.post);
```

ljharb replied:
Why is this better? The private method is robust; this can be mutated.

aduh95 replied:
Both versions can be mutated, it’s better because it’s less code.

ljharb replied:
We should be optimizing for less observable/mutable runtime behavior not less code.

ljharb replied:
Put another way, more declarative code is far better than less imperative code.

aduh95 replied:
I suppose it's a point of vue, one could argue we should write code that produces the least amount of JS objects to save some memory. Anyway, @EricWendel pick the version you prefer, my suggestion is not blocking (although, if you're not taking my suggestion, please use `FunctionPrototypeBind` instead of `.bind` line 6).

ErickWendel replied:
Thanks @aduh95 and @ljharb just changed it!

aduh95 commented on test/parallel/test-inspector-promises.js:
```diff
@@ -0,0 +1,63 @@
+const assert = require('assert');
+const inspector = require('inspector/promises');
+
+const {
+  basename
+} = require('path');
```
```suggestion
const {
  basename,
} = require('path');
```
or
```suggestion
const { basename } = require('path');
```

aduh95 commented on test/parallel/test-inspector-promises.js:
```diff
@@ -0,0 +1,63 @@
+
+    const {
+      callFrame: {
+        url
+      }
```
```suggestion
        url,
      },
```

aduh95 commented on test/parallel/test-inspector-promises.js:
```diff
@@ -0,0 +1,63 @@
+        url
+      }
+    } = profile.nodes.find(({
+      callFrame
```
```suggestion
      callFrame,
```

aduh95 commented:
This is adding a new core module requirable without the `node:` prefix and therefor is a potential breaking change (/cc @nodejs/tsc). It can land as a https://github.com/nodejs/node/labels/semver-minor if we add it to the following set: https://github.com/nodejs/node/blob/472edc775d683aed2d9ed39ca7cf381f3e7e3ce2/lib/internal/bootstrap/loaders.js#L124-L127

mcollina reviewed (approved):
lgtm

benjamingr commented:
@aduh95 I don't think it's a breaking change though? `inspector` is already a Node core module and this is adding a `/promises` suffix to it. I don't recall we considered that exposing a new module in the past.

(Additionally it I think it would be weird if you could import `inspector` but had to write `node:inspector/promises` for the promise version in terms of consistency)

aduh95 commented:
@benjamingr it is a breaking change:

```console
$ mkdir repro && cd repro
$ mkdir -p node_modules/inspector
$ echo 'module.exports = "user module";' > node_modules/inspector/promises.js
$ node -p 'require("inspector/promises")'
"user module"
```

ljharb commented:
node could ship a breaking change that all subpaths of any core module can never be required, which would free up these things so they're non-major in the future.

cjihrig commented:
> This is adding a new core module requirable without the `node:` prefix and therefor is a potential breaking change (/cc @nodejs/tsc). It can land as a [ semver-minor](https://github.com/nodejs/node/labels/semver-minor) PRs that contain new features and should be released in the next minor version. if we add it to the following set:
> 
> https://github.com/nodejs/node/blob/472edc775d683aed2d9ed39ca7cf381f3e7e3ce2/lib/internal/bootstrap/loaders.js#L124-L127

As much as I don't like the `node:` prefix only modules, why would we not do this at this point? I didn't want to implement prefix only modules for the test runner, but was outvoted... so why would we not take advantage of that moving forward?

ljharb commented:
for the same reasons not to do it with "test" - user confusion. That "x works normally and x/promises doesn't" is a much larger problem than "x works normally and y doesn't".

cjihrig commented:
@ljharb I personally agree with you, but since the TSC decided that we should go that route moving forward, I think we should stick to that. Otherwise, I think we should back out the requirement for `node:test` (especially because we control the `test` module on npm now), but that is another discussion.

benjamingr commented:
I think @ljharb 's suggestion about making sub-paths of core modules guarded is good and we should do that.

I do see the potential for someone to publish a new version of a userland [`inspector`](https://www.npmjs.com/package/inspector) and export a promises path that does something malicious. While that is a pretty theoretical concern (and the existing npm module doesn't export that path) - the fact it's a potential security risk does merit some caution.

ljharb commented:
@cjihrig that's not my understanding of the TSC vote - my understanding is that was a one-off decision just for test.

cjihrig commented:
Fair enough.

aduh95 commented:
> I do see the potential for someone to publish a new version of a userland [`inspector`](https://www.npmjs.com/package/inspector) and export a promises path that does something malicious. While that is a pretty theoretical concern (and the existing npm module doesn't export that path) - the fact it's a potential security risk does merit some caution.

FWIW that's only true if the user doesn't use the `node:` prefix.

benjamingr commented:
> FWIW that's only true if the user doesn't use the node: prefix.

Yes, we can support `node:` as a prefix anyway though I do feel strongly that if the core module is available without a prefix the /promises should be too.

khaosdoctor commented:
Leaving my two cents here as an user (hoping to contributing more 😅).

I think the use of `node:` prefix is very helpful especially to keep a standard, I don't know if there were other discussions about it in the past or in other issues (I'd love to check them out) but if there's a way to use it in the module it'd be great (assuming this wouldn't break the rest of things)

And on the `promises` suffix, haven't this been done on `fs` and `fs/promises`?

mhdawson commented:
> @cjihrig that's not my understanding of the TSC vote - my understanding is that was a one-off decision just for test.

While I can see the vote questions were specific, I had read more into it than a pure one off.  We should probably have a vote on general direction as one-offs don't make as much sense in my opinion. We should have a long direction that we agree to go in and then assume that is the default unless there is a specific/good reason to devicate.

mcollina commented:
What's the question for the TSC?

aduh95 commented:
> What's the question for the TSC?

@benjamingr correct me if I'm wrong, but I think the question is "should this land as semver-major and be importable/loadable without `node:` prefix, or as semver-minor under the `node:` prefix only?" And more broadly, do we want as TSC to have a rule that all new core modules should be prefix only.

tniessen commented:
> I thought about it, but the only function that makes sense using promises is the post function.

Could you elaborate on this? Are `connect()`, `connectToMainThread()`, `disconnect()` etc. required to be synchronous APIs for some technical reason?

It would also be nice not having to use `fs.writeSync` etc., e.g., by using async iterators to iterate over session events.

benjamingr commented:
> correct me if I'm wrong, but I think the question is "should this land as semver-major and be importable/loadable without node: prefix, or as semver-minor under the node: prefix only?" 

That is correct - should this land as semver-major (without the node: prefix, and semver-minor with the prefix) or as semver-minor (supporting both)? 

More broadly - I'd like for the TSC to either agree or disagree with "Adding /promises versions of existing modules is/isn't semver-major"

targos commented:
Why would you like the TSC to make the decision? Is there a disagreement between collaborators?

benjamingr commented:
@targos yes, I believe this is fine to add without the `node:` prefix and that doing so probably isn't a semver-major change. My reasoning is that this is what we've done in the past when adding `/promises` APIs.

@aduh95 's point, if I'm understanding it correctly which is also totally reasonable and valid is that this should (by default) be semver-major (see https://github.com/nodejs/node/pull/44250#issuecomment-1229263060 ) since the TSC has never officially decided to exempt new paths of existing modules from the requirement.

Additionally - @ljharb raised a different point of a potentially breaking (but positive) change: let's go ahead and block subpaths of built in modules. This is a very reasonable idea IMO that would itself be semver-major (right?) but would in turn ensure that future changes similar to this would never be breaking (even in theory). I'd like that suggestion to also have the TSC's blessing.

(I'd also like to emphasize that the TSC being asked to weigh in and vote is just because of process involving a new module, guarantees of existing module names and semverness - the discussion was civil and productive and I'm not sure I'd frame it as disagreement)

ljharb commented:
I'd love to see this outcome:
1. a breaking change lands that blocks require/import of subpaths of existing core modules, so that new subpaths can then be added in a semver-minor
1. a subpath of a core module should be accessible by the identical mechanisms that the top-level core module is
1. new top-level core modules continue to be accessible without the `node:` prefix

I would hope that only the third is potentially controversial.

aduh95 commented:
> My reasoning is that this is what we've done in the past when adding `/promises` APIs.

I don't think that's correct? AFAIK adding new core modules have always been semver-major, that's why promises API were not added as their own module, but instead as e.g. `require('dns').promises`.



> let's go ahead and block subpaths of built in modules. This is a very reasonable idea IMO that would itself be semver-major (right?) but would in turn ensure that future changes similar to this would never be breaking (even in theory). I'd like that suggestion to also have the TSC's blessing.

FWIW this doesn't have my blessing, that doesn't seem a reasonable way of tackling the problem when it's already solved by having a prefix. I personally would prefer to see the prefix-less deprecated. The prefix may not be perfect, but I find it a much better DX than having to know by heart which packages are built-in, which ones are from user-land.

italojs commented:
Since we use don't specify the file properly to require the module, why not move the `lib/inspector.js` for `lib/inspector/index.js` I was looking at the `requires` and I think it doesn't will be a breaking change

ErickWendel commented:
> Since we use don't specify the file properly to require the module, why not move the `lib/inspector.js` for `lib/inspector/index.js` I was looking at the `requires` and I think it doesn't will be a breaking change



> Just a new suggestion that is not related to the topic above: Since we use don't specify the file properly to require the module, why not move the `lib/inspector.js` for `lib/inspector/index.js` I was looking at the `requires` and I think it doesn't will be a breaking change

I think this is a valid change. However, I'm not sure about its impact of it. All files that are using it as lib/inspector will benefit from it although when using ECMAScript modules we specify it as the full path (eg lib/inspector.js) so all ESM will have to be changed to import it from the new path.

I'd create a new PR just to make this kind of change and make sure all tests are working fine after it.

khaosdoctor commented:
> @aduh95: I personally would prefer to see the prefix-less deprecated. The prefix may not be perfect, but I find it a much better DX than having to know by heart which packages are built-in, which ones are from user-land.

I'd like to see this happening too, actually I was pretty excited with the prefix because it would standardize the core module system a bit more over the others and allow for deeper changes without breaking the rest. Deprecating the original `inspector` for `node:inspector` as a major would be acceptable for me as the final change for users would be minimal

ljharb commented:
A runtime deprecation would be a massive problem for the ecosystem, since any transitive dependencies on core modules would bother users with no way for them to fix it.

We shouldn't be deprecating something that works fine - and has worked fine for over a decade - without incredibly compelling justifications, and "we wish it had always been this way" isn't compelling imo.

khaosdoctor commented:
That makes sense, my point is that I think we either should strive for both being `node:` like `node:inspector` and `node:inspector/promises` or none of them I, personally, like the first approach but I understand this can bring a bunch of problems for users as a breaking change (however, I wouldn't discard that in, for instance, a major version release in the future like Node 20 or so). 

So, I think the good solution, at least for me, now would be to have none being required and used just the way it is without a breaking change, but having `/promises` only accessible as `node:inspector/promises` and the root module both accessible as `node:inspector` and `inspector` kinda confuses a bit.

mhdawson commented:
This was discussed in the TSC meeting today. The concensus was this can land as SemVer Major under inspector/promises.  You can check the TSC minutes for the guidance we are going to PR in general.

ljharb commented:
@mhdawson was there any TSC decision on specifier consistency with `inspector`?

GeoffreyBooth commented:
> The consensus was this can land as SemVer Major under inspector/promises.

To be clear, I think the intent was to land as semver major shipping both `inspector/promises` and `node:inspector/promises`. As in, there will be no difference whether the prefix is included or not.

If we backport, it would probably be as the same, if people judge that to be safe. Per today’s conversation we’re leaning against backporting some depends-on-prefix variation like `node:inspector/promises` without `inspector/promises`.

ljharb commented:
Awesome, thanks. I still hope we land a major change that unlocks future subpath additions as minor.

tniessen commented:
It would be great to get a response to https://github.com/nodejs/node/pull/44250#issuecomment-1232709495 before this lands, even if the response is just "yes, the other functions have to be sync because of xyz".

aduh95 commented:
@ErickWendel Any change you would have time to address https://github.com/nodejs/node/pull/44250#issuecomment-1241720449 before the v19.x semver cutoff (2022-10-11)? And also, this needs a rebase.

ErickWendel commented:
> @ErickWendel Any change you would have time to address [#44250 (comment)](https://github.com/nodejs/node/pull/44250#issuecomment-1241720449) before the v19.x semver cutoff (2022-10-11)? And also, this needs a rebase.

hey sure! I'm travelling so I think I will only be able to work on this by the weekend

aduh95 commented:
Can you fix the commit message please?

tniessen commented:
> I actually was following the current implementation. In the actual code, it runs async and the only function that runs asynchronously is the `.post`
> 
> However, I think this makes sense to wait for the events and turn them all in async ops to make sure the order of execution will always work fine. I can add it in another PR if needed

I am not sure I understand. Doesn't this imply that `connect()` etc. should return promises?

ErickWendel commented:
> > I actually was following the current implementation. In the actual code, it runs async and the only function that runs asynchronously is the `.post`
> > However, I think this makes sense to wait for the events and turn them all in async ops to make sure the order of execution will always work fine. I can add it in another PR if needed
> 
> I am not sure I understand. Doesn't this imply that `connect()` etc. should return promises?

You meant all functions in inspector/promises should return promises? For me, it does make sense however it'll need to rewrite the implementation there for all of them. 

What do you suggest for this PR?

tniessen commented:
> What do you suggest for this PR?

I'm somewhat surprised that this is not marked as experimental but I guess the semver-major label is supposed to compensate for that.

The point of introducing `/promises` for any API is to provide a modern `Promise`-based API. If this is not experimental, and if returning `Promise`s later is going to be a breaking change (e.g., because users must then `await` the returned `Promise`), then another semver-major change will be necessary. I don't know whether that's a good idea or not, so I'm happy to leave that decision up to you and those who approve this PR.

aduh95 commented:
+1 to land it as experimental, we can always mark it as stable in a semver-minor.

ErickWendel commented:
> > What do you suggest for this PR?
> 
> I'm somewhat surprised that this is not marked as experimental but I guess the semver-major label is supposed to compensate for that.
> 
> The point of introducing `/promises` for any API is to provide a modern `Promise`-based API. If this is not experimental, and if returning `Promise`s later is going to be a breaking change (e.g., because users must then `await` the returned `Promise`), then another semver-major change will be necessary. I don't know whether that's a good idea or not, so I'm happy to leave that decision up to you and those who approve this PR.

I'm +1 too to land it as experimental.

Also +1 on landing only promised functions on this module. Gonna change it to land only the post function for now

jasnell commented on doc/api/inspector.md:
```diff
@@ -262,6 +377,82 @@ session.post('HeapProfiler.takeHeapSnapshot', null, (err, r) => {
+parameter usage.
+
+### `inspector.url()`
+
```
These sections should also have the yaml metadata

ErickWendel commented:
Hey, just a heads up. I've been crazy busy these last weeks but I didn't give up on this task. I noticed I'll have to rebase it again and there're a lot of conflicts to be fixed first. 

I'll do my best to finish this week, sorry about that

ErickWendel commented:
@tniessen I haven't found a way to get if the session is connected or not. 

In the code: it just checks whether the prop in the context has any value.  

```js
connect() {
    if (this[connectionSymbol])
      throw new ERR_INSPECTOR_ALREADY_CONNECTED('The inspector session');
    this[connectionSymbol] =
      new Connection((message) => this[onMessageSymbol](message));
  }
```

In this case, should I just annotate the function as async to return a Promise or change its implementation to wait for connection events and return it? 

That's the same for disconnect. In case I'd need to listen for those events, I haven't found them in the implementation. What would you suggest for this?

aduh95 commented:
@ErickWendel I suggest we land it as is, and to refine the API in follow up PRs: since we are landing this as experimental, we should try to first meet the deadline for v19.0.0 with the current implementation IMO (I mean, it still needs a rebase to fix the conflict, but otherwise LGTM).

aduh95 commented:
@ErickWendel the feature freeze date is today, do you think you could quickly resolve the git conflict here so it can land?

ErickWendel commented:
> @ErickWendel the feature freeze date is today, do you think you could quickly resolve the git conflict here so it can land?

Sure! I'll do it now! To land it as experimental, should I change anything in the code?

aduh95 commented:
Please add https://github.com/nodejs/node/blob/39998c9cf403b5acfa21d2e442d733e8160af4fd/doc/api/assert.md#L2098 in the docs. We could also add a runtime warning, but I don't think it's necessary.

ErickWendel commented:
@aduh95 done rebasing it! Gonna keep improving it in the next PRs

aduh95 commented on doc/api/inspector.md:
```diff
@@ -11,92 +11,211 @@ inspector.
 
-## `inspector.close()`
+## Promises API
+
+<!--introduced_in=REPLACEME-->
```
```suggestion
<!-- YAML
added: REPLACEME
-->
```

aduh95 replied:
```suggestion
```

aduh95 commented:
It looks like this PR breaks `test.parallel/test-code-cache` when compiled without crypto/ without i18n support:

```
node:inspector:23
  throw new ERR_INSPECTOR_NOT_AVAILABLE();
  ^

Error [ERR_INSPECTOR_NOT_AVAILABLE]: Inspector is not available
    at new NodeError (node:internal/errors:393:5)
    at node:inspector:23:9
    at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/loaders:331:7)
    at nativeModuleRequire (node:internal/bootstrap/loaders:362:14)
    at node:inspector/promises:3:19
    at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/loaders:331:7)
    at BuiltinModule.compileForPublicLoader (node:internal/bootstrap/loaders:269:10)
    at loadBuiltinModule (node:internal/modules/cjs/helpers:50:9)
    at Module._load (node:internal/modules/cjs/loader:833:20)
    at Module.require (node:internal/modules/cjs/loader:1061:19) {
  code: 'ERR_INSPECTOR_NOT_AVAILABLE'
}

Node.js v19.0.0-pre
```

@ErickWendel Can you have a look?

ErickWendel commented:
> It looks like this PR breaks `test.parallel/test-code-cache` when compiled without crypto/ without i18n support:
> 
> ```
> node:inspector:23
>   throw new ERR_INSPECTOR_NOT_AVAILABLE();
>   ^
> 
> Error [ERR_INSPECTOR_NOT_AVAILABLE]: Inspector is not available
>     at new NodeError (node:internal/errors:393:5)
>     at node:inspector:23:9
>     at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/loaders:331:7)
>     at nativeModuleRequire (node:internal/bootstrap/loaders:362:14)
>     at node:inspector/promises:3:19
>     at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/loaders:331:7)
>     at BuiltinModule.compileForPublicLoader (node:internal/bootstrap/loaders:269:10)
>     at loadBuiltinModule (node:internal/modules/cjs/helpers:50:9)
>     at Module._load (node:internal/modules/cjs/loader:833:20)
>     at Module.require (node:internal/modules/cjs/loader:1061:19) {
>   code: 'ERR_INSPECTOR_NOT_AVAILABLE'
> }
> 
> Node.js v19.0.0-pre
> ```
> 
> @ErickWendel Can you have a look?

checking it now. How could I test without crypto and i18n?

aduh95 commented:
By using `./configure --without-ssl --without-intl`.

ErickWendel commented:
> By using `./configure --without-ssl --without-intl`.

Perfect, waiting to compile it here. I'd take a while

ErickWendel commented:
@aduh95 I compiled it using: 

```sh
./configure --node-builtin-modules-path $(pwd) --without-ssl --without-intl
make -j8
```

However, when I run 

```sh
./node test/parallel/test-inspector-promises.js
```
I get 
```
1..0 # Skipped: V8 inspector is disabled
```

I'm not sure if there is any config to skip this test on Node. Do you know what could it be?

Running the `test-code-cache` I get:

```sh
./node test/parallel/test-code-cache.js 
node:internal/modules/cjs/loader:998
  throw err;
  ^
Error: Cannot find module 'internal/test/binding'
```

aduh95 commented:
You need to use the Python runner:

```sh
python3 tools/test.py test/parallel/test-code-cache
```

ErickWendel commented:
> You need to use the Python runner:
> 
> ```shell
> python3 tools/test.py test/parallel/test-code-cache
> ```

working on this now, but seeing the CI, it's not failing anymore (not sure why oO)

ErickWendel commented:
Done! There's a rule for `inspector` on `src/node_builtins` that must be added to `inspector/promises` as well

aduh95 commented:
@ErickWendel can you rebase please? Merge commits tend to break our tooling.

ErickWendel commented:
> @ErickWendel can you rebase please? Merge commits tend to break our tooling.

Oh sure, I merged it from the interface as it was just a single line. Gonna do it now

ErickWendel commented:
just rebased it. Gonna have to leave know but lmk if there is anything pending so I can do it later night

ovflowd commented on doc/api/inspector.md:
````diff
@@ -11,94 +11,209 @@ inspector.
 It can be accessed using:
 
-```js
+```mjs
````
Doc-wise, this should be somewhere else, And marked with `> Stability 1 Experimental.

E.g.:

````
## Promises

<!-- introduced_in=v19.0.0 -->

> Stability: 1 - Experimental

<!-- source_link=lib/inspector/promises.js

[...] Description.
```

IMHO this is a doc-blocker.

aduh95 replied:
It is marked as experimental, see https://github.com/nodejs/node/pull/44250/files#diff-d6da7e100224e744520fde97ca99f743ad9eafc83d50ed92ed2b7870f0589f68R34

`<!-- introduced_in=v19.0.0 -->` would not be appropriate here, those are to reference the first version that included the markdown document in the docs. This PR does contain a YAML comment block that does the same thing: https://github.com/nodejs/node/pull/44250/files#diff-d6da7e100224e744520fde97ca99f743ad9eafc83d50ed92ed2b7870f0589f68R37.

I disagree this is a doc blocker, also doc can always be fixed in a semver patch, while the module can only land in a semver-major.

ovflowd replied:
Not saying this is a PR blocker, but felt for me like a doc blocker, if you disagree that's fine, I don't want to block this PR too much 👀

ErickWendel replied:
I didn't get whether i should change it or not haha. Is it ready to be landed? cc @aduh95 @ovflowd

ErickWendel replied:
Perfect! @aduh95. 

for the next PR @ovflowd I didn't get what you suggested to change. I plan to:

change connect and disconnect functions to be async (this might lead me to change something in cpp to get those events (no idea how to do it haha)) and what you suggested adding here.

ovflowd replied:
Hmm, @ErickWendel not sure, but I was just making suggestions to the API docs you were writing xD

But it's all fine and this landed 🎉

ovflowd reviewed (changes requested):
Just adding a blocker for docs 👀 as it doesn't follow the correct standard.
